README.md
# Flutter Circle Chart

A Flutter package to create a Circle Chart with multiple drawing types. It will be useful for your awesome app.

![Demo](./demo.gif)

### Example

```dart
import 'dart:math';
import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:flutter_circle_chart/flutter_circle_chart.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        backgroundColor: Color(0xff4C2882),
        appBar: AppBar(
          title: Text('Flutter Circle Chart'),
        ),
        body: SingleChildScrollView(
          child: Column(
            children: List.generate(
              CircleChartType.values.length,
              (index) => Padding(
                padding: const EdgeInsets.all(8.0),
                child: Column(
                  children: [
                    Padding(
                      padding: const EdgeInsets.all(8.0),
                      child: Text(
                        '${CircleChartType.values[index]}',
                        style: TextStyle(
                          fontSize: 16,
                          color: Colors.white,
                          fontWeight: FontWeight.bold,
                        ),
                      ),
                    ),
                    CircleChart(
                      chartType: CircleChartType.values[index],
                      items: List.generate(
                        3,
                        (index) => CircleChartItemData(
                          color: randomColor(),
                          value: 100 + Random.secure().nextDouble() * 1000,
                          name: 'Lorem Ipsum $index',
                          description:
                              'Lorem Ipsum $index không phải chỉ là một đoạn văn bản ngẫu nhiên.',
                        ),
                      ),
                    ),
                  ],
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}

Color randomColor() {
  var g = math.Random.secure().nextInt(255);
  var b = math.Random.secure().nextInt(255);
  var r = math.Random.secure().nextInt(255);
  return Color.fromARGB(255, r, g, b);
}
```

### Development environment

```
[✓] Flutter (Channel stable, 2.0.5, on macOS 11.2.3 20D91 darwin-x64, locale en-VN)
    • Flutter version 2.0.5
    • Framework revision adc687823a (11 days ago), 2021-04-16 09:40:20 -0700
    • Engine revision b09f014e96
    • Dart version 2.12.3

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Platform android-30, build-tools 30.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode_12.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
```

CHANGELOG.md
## [1.0.0] - 27/06/2021

* First release version.


lib/math/math.dart
import 'dart:math' as math;

export 'affineTransform.dart';
export 'objectives.dart';

double sqrt(double a) => math.sqrt(a);
double pow(double a, double b) => math.pow(a, b).toDouble();
double abs(double a) => a < 0 ? -a : a;
int round(double a) => a.round();
double atan(double a) => math.atan(a);
double cos(double a) => math.cos(a);
double sin(double a) => math.sin(a);
const pi = math.pi;

const kDegToRad = math.pi / 180.0;
const kRadToDeg = 180.0 / math.pi;

double toRad(double degree) => degree * kDegToRad;

double factorial(double x) {
  if (x < 3) {
    if (x < 2)
      return 1;
    else
      return 2;
  } else {
    double res = 1;
    for (double i = 2; i <= x; i++) {
      res *= i;
    }
    return res;
  }
}

const int sign_mask = 0x80000000;
const double b = 0.596227;
double atan2(double y, double x) {
  // Extract the sign bits
  int ux_s = sign_mask & x.round();
  int uy_s = sign_mask & y.round();

  // Determine the quadrant offset
  double q = ((~ux_s & uy_s) >> 29 | ux_s >> 30).toDouble();

  // Calculate the arctangent in the first quadrant
  double bxy_a = abs(b * x * y);
  double num = bxy_a + y * y;
  double atan_1q = num / (x * x + bxy_a + num);

// Translate it to the proper quadrant
  int uatan_2q = (ux_s ^ uy_s) | atan_1q.round();
  return q + uatan_2q.toDouble();
}


lib/math/objectives.dart
class Vector {
  double B;
  double A;

  Vector([this.A = 0, this.B = 0]);
}

class Point {
  double x;
  double y;

  Point([this.x = 0.0, this.y = 0.0]);

  static Point zero = Point(0.0, 0.0);

  factory Point.fromMap(Map<String, dynamic> map) {
    return new Point(
      map['x'] as double,
      map['y'] as double,
    );
  }

  Map<String, dynamic> toMap() {
    // ignore: unnecessary_cast
    return {
      'x': this.x,
      'y': this.y,
    } as Map<String, dynamic>;
  }

  @override
  String toString() {
    return toMap().toString();
  }
}

/// Ax + By + C = 0
class LineEquation {
  double A = 0;
  double B = 0;
  double C = 0;

  bool isNull() {
    bool ans = false;
    if (A == 0 && B == 0 && C == 0)
      ans = true;
    else if (A == 0 && B == 0)
      ans = true;
    else
      ans = false;
    return ans;
  }
}

class Polygon {
  final List<Point> vertices;

  Polygon(this.vertices) : assert(vertices != null);

  Point gravityCenter() {
    if (vertices.isEmpty) return Point(0, 0);
    double x = 0.0, y = 0.0;
    for (int i = 0; i < count; i++) {
      x += vertices[i].x;
      y += vertices[i].y;
    }
    return Point(x / count, y / count);
  }

  int get count => vertices.length;
}

class Circle {
  final Point center;
  final double R;

  Circle(this.center, this.R);
}


lib/flutter_circle_chart.dart
import 'dart:math' as math;
import 'dart:ui' as ui;

import 'package:flutter/material.dart';

// const _kDegToRad = math.pi / 180.0;
// const _kRadToDeg = 180.0 / math.pi;
const _kTotalRad = 2 * math.pi;
const _kSelectedItemFactor = 1.3;

/// The data is used to draw the [CircleChart]
/// Each item has [CircleChartItemData.value] and the [CircleChart] will calculate total value itself and display as a label.
class CircleChartItemData {
  /// [color] : color to draw.
  final Color color;

  /// [value] : value to calculate its part of the [CircleChart].
  final double value;

  /// [name] : is used to call each item.
  final String name;

  /// [description] : Item description is used to describe more details each item.
  final String description;

  /// Constructor
  CircleChartItemData({
    required this.color,
    required this.value,
    required this.name,
    required this.description,
  });
}

/// Drawing types of [CircleChart]
/// [CircleChartType.solid] draws items in solid colors.
/// [CircleChartType.gradient] draws items in gradients from their [CircleChartItemData.color] darker to their origin color.
/// [CircleChartType.bracelet] like as gradient, but it come from the center, make the chart like a bracelet.
/// [CircleChartType.dots] draws items in dots with their [CircleChartItemData.color].
enum CircleChartType {
  solid,
  gradient,
  bracelet,
  dots,
}

/// Help you create a Circle Chart that is used to display some kinds of reports.
class CircleChart extends StatefulWidget {
  /// [backgroundColor] : the background of whole widget.
  final Color backgroundColor;

  /// [borderRadius] : the border radius of whole widget.
  final Radius borderRadius;

  /// [padding] : the padding of whole widget.
  final EdgeInsets padding;

  /// [duration] : use for the animation, triggered whenever the [items] changes or not by [animationOnItemsChanged].
  final Duration duration;
  final bool animationOnItemsChanged;

  /// [chartRadius] : radius to draw the circle.
  final double chartRadius;

  /// [chartType] : drawing type. See all [CircleChartType].
  final CircleChartType chartType;

  /// [chartStrokeWidth] : stroke width of the item value circle.
  final double chartStrokeWidth;

  /// [chartCircleBackgroundStrokeWidth] : stroke width of the dark circle behind the item value circle.
  final double chartCircleBackgroundStrokeWidth;

  /// See [labelTextStyle]
  final Radius labelBorderRadius;

  /// See [labelTextStyle]
  final EdgeInsets labelPadding;

  /// The label is used to show the total value and selected item value to focus on. [labelTextStyle] is used for Text. [labelBorderRadius] & [labelPadding] are used for the label background.
  final TextStyle labelTextStyle;

  /// [items] : is the data to draw.
  final List<CircleChartItemData> items;

  /// [itemPadding] : is the padding of each item placed at the right.
  final EdgeInsets itemPadding;

  /// [itemTextStyle] : is the text style of each item placed at the right.
  final TextStyle itemTextStyle;

  /// [itemDescriptionTextStyle] : is the text style of description label of each item placed at the bottom of item label.
  final TextStyle itemDescriptionTextStyle;

  /// Create an CircleChart. It has a default style.
  const CircleChart({
    Key? key,
    this.backgroundColor: const Color(0xff32074e),
    this.borderRadius: const Radius.circular(8),
    this.padding: const EdgeInsets.all(12),
    this.chartRadius: 60,
    this.labelPadding: const EdgeInsets.all(12),
    this.labelBorderRadius: const Radius.circular(8),
    this.labelTextStyle: const TextStyle(
        color: Colors.white, fontSize: 16, fontWeight: FontWeight.bold),
    this.itemTextStyle: const TextStyle(
      color: Colors.white,
      fontSize: 12,
    ),
    this.itemDescriptionTextStyle: const TextStyle(
      color: Colors.white60,
      fontSize: 10,
    ),
    required this.items,
    this.chartType: CircleChartType.dots,
    this.chartStrokeWidth: 15,
    this.chartCircleBackgroundStrokeWidth: 20,
    this.itemPadding: const EdgeInsets.symmetric(horizontal: 12, vertical: 8),
    this.duration: const Duration(milliseconds: 600),
    this.animationOnItemsChanged: true,
  }) : super(key: key);

  @override
  _CircleChartState createState() => _CircleChartState();
}

class _CircleChartState extends State<CircleChart>
    with SingleTickerProviderStateMixin {
  CircleChartItemData? _selectedItem;
  late AnimationController _animationController;

  @override
  void initState() {
    _animationController =
        AnimationController(vsync: this, duration: widget.duration);
    _animationController.addListener(() {
      setState(() {});
    });
    _startAnimation();
    super.initState();
  }

  @override
  void didUpdateWidget(covariant CircleChart oldWidget) {
    if (widget.animationOnItemsChanged) {
      if (oldWidget.items != widget.items) {
        _selectedItem = null;
        _startAnimation();
      }
    }
    super.didUpdateWidget(oldWidget);
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: widget.padding,
      decoration: BoxDecoration(
        color: widget.backgroundColor,
        borderRadius: BorderRadius.all(widget.borderRadius),
      ),
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          CustomPaint(
            painter: _CircleChartPainter(
              items: widget.items,
              chartType: widget.chartType,
              chartRadius: widget.chartRadius,
              strokeWidth: widget.chartStrokeWidth,
              circleStrokeWidth: widget.chartCircleBackgroundStrokeWidth,
              animationValue: CurveTween(curve: Curves.easeInOutCirc)
                  .evaluate(_animationController),
              selectedItem: _selectedItem,
            ),
            size: Size(widget.chartRadius * 2, widget.chartRadius * 2),
          ),
          Expanded(
              child: Padding(
            padding: EdgeInsets.only(left: widget.itemPadding.left),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.stretch,
              children: [
                Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    Container(
                      alignment: Alignment.center,
                      padding: widget.labelPadding,
                      decoration: BoxDecoration(
                        borderRadius:
                            BorderRadius.all(widget.labelBorderRadius),
                        color: _selectedItem?.color ?? Colors.black54,
                      ),
                      child: Text(
                        '${(_selectedItem != null ? _selectedItem!.value : _calculateTotalValue(widget.items)).toStringAsFixed(2)}',
                        style: widget.labelTextStyle,
                      ),
                    ),
                  ],
                ),
                ...widget.items.map((e) => Padding(
                      padding: EdgeInsets.only(top: widget.itemPadding.top),
                      child: _buildItem(e),
                    )),
              ],
            ),
          )),
        ],
      ),
    );
  }

  Widget _buildItem(CircleChartItemData item) {
    return ElevatedButton(
      onPressed: () {
        setState(() {
          if (_selectedItem != item || _selectedItem == null) {
            _selectedItem = item;
          } else {
            _selectedItem = null;
          }
        });
      },
      style: ButtonStyle(
        shape: MaterialStateProperty.all<OutlinedBorder>(RoundedRectangleBorder(
          borderRadius: BorderRadius.all(widget.borderRadius),
        )),
        padding: MaterialStateProperty.all<EdgeInsets>(widget.itemPadding),
        elevation: MaterialStateProperty.all<double>(0.0),
        tapTargetSize: MaterialTapTargetSize.shrinkWrap,
        minimumSize: MaterialStateProperty.all<Size>(Size.zero),
        backgroundColor: MaterialStateProperty.all<Color>(Colors.transparent),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Container(
                width: widget.itemTextStyle.fontSize,
                height: widget.itemTextStyle.fontSize,
                decoration: ShapeDecoration(
                  color: item.color,
                  shape: CircleBorder(),
                ),
              ),
              Expanded(
                child: Padding(
                  padding: EdgeInsets.only(left: widget.itemPadding.left),
                  child: Text(
                    '${item.value.toStringAsFixed(2)} ${item.name}',
                    style: widget.itemTextStyle,
                  ),
                ),
              ),
            ],
          ),
          if (_selectedItem == item)
            Padding(
              padding: EdgeInsets.only(top: widget.itemPadding.top),
              child: Text(
                item.description,
                style: widget.itemDescriptionTextStyle,
              ),
            ),
        ],
      ),
    );
  }

  void _startAnimation() {
    _animationController.forward(from: 0.1);
  }
}

class _CircleChartPainter extends CustomPainter {
  final double chartRadius;
  final List<CircleChartItemData> items;
  final CircleChartType chartType;
  final double strokeWidth;
  final double circleStrokeWidth;
  final double animationValue;
  final CircleChartItemData? selectedItem;

  const _CircleChartPainter({
    required this.chartRadius,
    required this.items,
    required this.chartType,
    required this.strokeWidth,
    required this.circleStrokeWidth,
    this.animationValue: 1.0,
    this.selectedItem,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final center = size.center(Offset.zero);
    final radius = chartRadius - circleStrokeWidth * 0.5;
    final totalValue = _calculateTotalValue(items);
    // final count = items.length;
    final totalRad = animationValue * _kTotalRad;
    double startAngle = 0;

    /// draw circle background
    canvas.drawCircle(
        center,
        radius,
        Paint()
          ..color = Colors.black54
          ..strokeWidth = circleStrokeWidth
          ..strokeCap = StrokeCap.round
          ..strokeJoin = StrokeJoin.round
          ..style = PaintingStyle.stroke);

    /// draw arc items
    switch (chartType) {
      case CircleChartType.bracelet:
        startAngle = (totalRad * items.first.value / totalValue);
        for (var i = 0; i < items.length; ++i) {
          var item = items[i];
          var sweepAngle = (totalRad * item.value / totalValue);
          canvas.drawArc(
              Rect.fromCircle(center: center, radius: radius),
              startAngle,
              sweepAngle,
              false,
              Paint()
                ..strokeWidth = strokeWidth *
                    (selectedItem == item ? _kSelectedItemFactor : 1)
                ..strokeCap = StrokeCap.round
                ..strokeJoin = StrokeJoin.round
                ..style = PaintingStyle.stroke
                ..shader = ui.Gradient.radial(
                  center,
                  radius,
                  [
                    item.color.darker,
                    item.color,
                  ],
                  [0.9, 1.0],
                )
              //
              );
          startAngle += sweepAngle;
        }
        break;
      case CircleChartType.gradient:
        startAngle = (totalRad * items.first.value / totalValue);
        for (var i = 1; i < items.length; ++i) {
          var item = items[i];
          var sweepAngle = totalRad * item.value / totalValue;
          canvas.drawArc(
              Rect.fromCircle(center: center, radius: radius),
              startAngle,
              sweepAngle,
              false,
              Paint()
                ..strokeWidth = strokeWidth *
                    (selectedItem == item ? _kSelectedItemFactor : 1)
                ..strokeCap = StrokeCap.round
                ..strokeJoin = StrokeJoin.round
                ..style = PaintingStyle.stroke
                ..shader = ui.Gradient.sweep(
                  center,
                  [
                    item.color,
                    item.color.darker,
                  ],
                  [0.0, 1.0],
                  TileMode.clamp,
                  startAngle,
                  startAngle + sweepAngle,
                )
              //
              );
          startAngle += sweepAngle;
        }
        startAngle = 0;
        var item = items.first;
        var sweepAngle = ((totalRad) * item.value / totalValue);
        canvas.drawArc(
            Rect.fromCircle(center: center, radius: radius),
            startAngle,
            sweepAngle,
            false,
            Paint()
              ..strokeWidth = strokeWidth *
                  (selectedItem == item ? _kSelectedItemFactor : 1)
              ..strokeCap = StrokeCap.round
              ..strokeJoin = StrokeJoin.round
              ..style = PaintingStyle.stroke
              ..color = item.color
            //
            );
        break;
      case CircleChartType.dots:
        var totalDots = ((totalRad * radius) / (strokeWidth * 1.5)).round();
        totalDots = totalDots + 3 - totalDots % 3;
        // print('CircleChartPainter.paint: totalDots=$totalDots');
        var startPoint = Offset(center.dx + radius, center.dy);
        Offset c = startPoint;
        // var totalDots = 18;
        var totalCount = 0;
        for (var item in items) {
          var percent = item.value / totalValue;
          late int count;
          if (item == items.last) {
            count = totalDots - totalCount;
          } else {
            count = (percent * totalDots).round();
          }

          // print(
          //     'CircleChartPainter.paint: ${item.value.toStringAsFixed(2)} ~ $percent ~ $count dots');
          final da = totalRad / totalDots;
          for (var i = 0; i < count; ++i) {
            canvas.drawCircle(
                c,
                strokeWidth *
                    0.5 *
                    (selectedItem == item ? _kSelectedItemFactor : 1),
                Paint()
                  ..color = item.color
                  ..style = PaintingStyle.fill
                //
                );
            c = c.rotate(center, da);
          }
          totalCount += count;
        }
        break;
      case CircleChartType.solid:
      default:
        for (var item in items) {
          var sweepAngle = (totalRad * item.value / totalValue);
          canvas.drawArc(
              Rect.fromCircle(center: center, radius: radius),
              startAngle,
              sweepAngle,
              false,
              Paint()
                ..color = item.color
                ..strokeWidth = strokeWidth *
                    (selectedItem == item ? _kSelectedItemFactor : 1)
                ..strokeCap = StrokeCap.round
                ..strokeJoin = StrokeJoin.round
                ..style = PaintingStyle.stroke);
          startAngle += sweepAngle;
        }
        break;
    }
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return true;
  }
}

/// calculate total value
double _calculateTotalValue(List<CircleChartItemData> items) {
  double a = 0;
  for (var item in items) {
    a += item.value;
  }
  return a;
}

extension _ColorEx on Color {
  /// build a color darker from its origin.
  get darker => Color.alphaBlend(this.withOpacity(0.5), Colors.black);
}

extension _OffsetEx on Offset {
  /// rotate a point by a [alphaRad] around a circle that has the [center] point.
  Offset rotate(Offset center, double alphaRad) {
    if (alphaRad == double.infinity || alphaRad == double.negativeInfinity)
      return this;
    // print('rotate: $center, $alphaRad');
    var matrix = Matrix4.identity()..translate(-center.dx, -center.dy);
    var p = MatrixUtils.transformPoint(matrix, this);
    matrix = Matrix4.identity()..rotateZ(alphaRad);
    p = MatrixUtils.transformPoint(matrix, p);
    matrix = Matrix4.identity()..translate(center.dx, center.dy);
    p = MatrixUtils.transformPoint(matrix, p);
    return p;
  }
}


lib/math/affineTransform.dart
import 'math.dart';

List<List<double>> get _emptyMatrix => [
      [0.0, 0.0, 0.0],
      [0.0, 0.0, 0.0],
      [0.0, 0.0, 0.0]
    ];
List<List<double>> get _identityMatrix => [
      [1.0, 0.0, 0.0],
      [0.0, 1.0, 0.0],
      [0.0, 0.0, 1.0]
    ];

_copyMatrix(List<List<double>> a, List<List<double>> b) {
  for (var i = 0; i < b.length; ++i) {
    for (var j = 0; j < b[i].length; ++j) {
      a[i][j] = b[i][j];
    }
  }
}

class AffineTransform {
  late List<List<double>> M;

  AffineTransform(this.M);

  AffineTransform.identity() {
    M = _identityMatrix;
  }

  void identity() {
    M = _identityMatrix;
  }

  void _multiply(List<List<double>> A) {
    var res = _emptyMatrix;
    for (int i = 0; i < 3; i++) {
      for (int j = 0; j < 3; j++) {
        res[i][j] = 0.0;
        for (int k = 0; k < 3; k++) res[i][j] += A[i][k] * M[k][j];
      }
    }
    _copyMatrix(M, res);
  }

  void apply(Point p) {
    double x = p.x, y = p.y;
    p.x = M[0][0] * x + M[0][1] * y + M[0][2];
    p.y = M[1][0] * x + M[1][1] * y + M[1][2];
  }

  void applyPolygon(Polygon pn) {
    for (int i = 0; i < pn.count; i++) {
      apply(pn.vertices[i]);
    }
  }

  void applyPoints(List<Point> points) {
    for (int i = 0; i < points.length; i++) {
      apply(points[i]);
    }
  }

  void translate(double tx, double ty) {
    var mt = _identityMatrix;
    mt[0][2] = tx;
    mt[1][2] = ty;
    _multiply(mt);
  }

  void forward(double distance, double angleRad) {
    double dx = distance * cos(angleRad), dy = distance * sin(angleRad);
    translate(dx, dy);
  }

  void rotateOrigin(double alpha) // tam la goc toa do
  {
    var mr = _identityMatrix;
    alpha *= kDegToRad;
    mr[0][0] = mr[1][1] = cos(alpha);
    mr[0][1] = -sin(alpha);
    mr[1][0] = sin(alpha);
    _multiply(mr);
  }

  void rotate(double alpha, Point center) // tam bat ki
  {
    translate(-center.x, -center.y);
    rotateOrigin(alpha);
    translate(center.x, center.y);
  }

  void scaleOrigin(double sx, double sy) // tam la goc toa do
  {
    var msc = _identityMatrix;
    msc[0][0] = sx;
    msc[1][1] = sy;
    _multiply(msc);
  }

  void shearOrigin(double hx, double hy) // tam la goc toa do
  {
    var msh = _identityMatrix;
    msh[0][1] = hx;
    msh[1][0] = hy;
    _multiply(msh);
  }

  void scale(double sx, double sy, Point center) // tam bat ki
  {
    translate(-center.x, -center.y);
    scaleOrigin(sx, sy);
    translate(center.x, center.y);
  }

  void shear(double hx, double hy, Point center) // tam bat ki
  {
    translate(-center.x, -center.y);
    shearOrigin(hx, hy);
    translate(center.x, center.y);
  }
}


example/lib/main.dart
import 'dart:math';
import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:flutter_circle_chart/flutter_circle_chart.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        backgroundColor: Color(0xff4C2882),
        appBar: AppBar(
          title: Text('Flutter Circle Chart'),
        ),
        body: SingleChildScrollView(
          child: Column(
            children: List.generate(
              CircleChartType.values.length,
              (index) => Padding(
                padding: const EdgeInsets.all(8.0),
                child: Column(
                  children: [
                    Padding(
                      padding: const EdgeInsets.all(8.0),
                      child: Text(
                        '${CircleChartType.values[index]}',
                        style: TextStyle(
                          fontSize: 16,
                          color: Colors.white,
                          fontWeight: FontWeight.bold,
                        ),
                      ),
                    ),
                    CircleChart(
                      chartType: CircleChartType.values[index],
                      items: List.generate(
                        3,
                        (index) => CircleChartItemData(
                          color: randomColor(),
                          value: 100 + Random.secure().nextDouble() * 1000,
                          name: 'Lorem Ipsum $index',
                          description:
                              'Lorem Ipsum $index không phải chỉ là một đoạn văn bản ngẫu nhiên.',
                        ),
                      ),
                    ),
                  ],
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}

Color randomColor() {
  var g = math.Random.secure().nextInt(255);
  var b = math.Random.secure().nextInt(255);
  var r = math.Random.secure().nextInt(255);
  return Color.fromARGB(255, r, g, b);
}
